### PR TITLE
Add AverageIndicator

### PR DESCRIPTION
**Pull request by Cypher01**
Add `AverageIndicator`, coming from my spin-off project [ta4j-extended](https://github.com/Cypher01/ta4j-extended)

### CODE REVIEW

**Comment by sgflt** [comment]
Isn't it just SMAIndicator?

**Comment by sgflt** [comment]
Ah I see. Average of current state of multiple indicators.

File: ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.List;
+
+/**
+ * Indicator to calculate the average of a list of other indicators.
+ */
+public class AverageIndicator extends CachedIndicator<Num> {
+    private final List<Indicator<Num>> indicators;
+
+    public AverageIndicator(Indicator<Num>... indicators) {
+        super(indicators[0]);
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java]: Please perform some input validation, at minimum the null and empty list cases.

File: ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+        this.indicators = Arrays.asList(indicators);
+    }
+
+    @Override
+    protected Num calculate(int index) {
+        Num value = getBarSeries().numFactory().zero();
+
+        for (Indicator<Num> indicator : indicators) {
+            value = value.plus(indicator.getValue(index));
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java:72]: What if the user instantiates an `AverageIndicator` with a list of indicators that happens to contain nulls?

File: ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        return value.dividedBy(getBarSeries().numFactory().numOf(indicators.size()));
+    }
+
+    @Override
+    public int getUnstableBars() {
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java]: `calculate` should invoke this function early on to check the current `index` against the unstable bars value. If within the unstable period we should return `NaN`, otherwise proceed normally.

File: ta4j-core/src/test/java/org/ta4j/core/indicators/helpers/AverageIndicatorTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Test
+    public void indicatorShouldRetrieveAverageOfOneAndTwoAndThree() {
+        AverageIndicator averageIndicator = new AverageIndicator(one, two, three);
+
+        for (int i = 1; i < 10; i++) {
+            assertNumEquals(2, averageIndicator.getValue(i));
+        }
+    }
```
Comment (TheCookieLab) [inline ta4j-core/src/test/java/org/ta4j/core/indicators/helpers/AverageIndicatorTest.java:82]: Please add tests to exercise the unstable bars scenarios.

**Review by TheCookieLab — CHANGES_REQUESTED** [review]

**Comment by TheCookieLab** [comment]
Thank you for contributing.

File: ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java (hunk 3 above)
Comment (Cypher01) [reply ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java]: Are you sure? Unstable bars sounds like the result is potentially not accurate, but not that it should be `NaN`. If I compare this with e.g. `SMAIndicator`, this one can also calculate the moving average from `index` 0. If I would feed two `SMAIndicator`'s of e.g. `barCount` 10 and 20 into `AverageIndicator`, it could calculate the average from `index` 0, but the value would not be accurate until `index` 20.

Comment (TheCookieLab) [reply ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java]: Why would one want to execute trades (particularly use cases where there's real money on the line) on potentially inaccurate signals? If the end user application is simply analyzing for "fun" or other low/no stakes scenarios then perhaps false positives/negatives are no big deal, but it's a very different story when real money is involved. Since there's no distinction between "play/test" and "serious/production" indicators, all indicators must be the latter by default.

The [emerging convention](https://github.com/ta4j/ta4j/issues/947) is for indicators to return `NaN` within their unstable period to clearly delineate between values that can be trusted (and tradeable) from those that are not. There's already indicators that do this so your `calculate` addition loop logic has to handle this anyway (should be a test case).

Bottom line is you'll need to do _something_ to handle the unstable period scenario. It doesn't necessarily have to be `NaN`, but that is the prevailing convention. Whatever it is, it must unambiguously preserve signal integrity.

Comment (Cypher01) [reply ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java]: The conversation is interesting. As you pointed out, "all indicators must be the latter by default". It's at least not consistent yet. And as I already pointed out, the term unstable bars suggests it is unstable, not that there is no data available.

But OK, I'll implement the changes.

File: ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java (hunk 1 above)
Comment (Cypher01) [reply ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java]: How should this look like? `super()` and `this()` must be first statement in a constructor. This implementation would anyway through an `ArrayIndexOutOfBoundsException`.

Either way, I implemented another constructor accepting a list of indicators and calling `super()` with `getFirst()`. But the situation is the same, just in this other case it will through a `NoSuchElementException`.

A manual validation would do the same, just the runtime exception would be a different one.
```
if (indicators == null || indicators.isEmpty()) {
    throw new IllegalArgumentException("...");
}
```

File: ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java (hunk 3 above)
Comment (sgflt) [reply ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java]: _is for indicators to return NaN within their unstable period to clearly delineate between values that can be trusted (and tradeable) from those that are not_ 

So do we need **_isStable_** method if instability is represented by NaN? https://github.com/ta4j/ta4j/issues/1141

Comment (TheCookieLab) [reply ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java]: Yes, converging on a stability identification and propagation solution will be a high priority item on the road map. I like the `isStable` approach right now but am open to alternatives.

**Review by TheCookieLab — APPROVED** [review]

File: ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java (hunk 1 above)
Comment (Cypher01) [reply ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/AverageIndicator.java]: You obviously found a nice solution, thx! :+1:

**Comment by TheCookieLab** [comment]
Alright barring any additional comments I'll be merging this shortly.

**Review by TheCookieLab — DISMISSED** [review]
